File: abtesting/src/main/java/fr/o80/lib/abtesting/ABPercentConfig.java
package fr.o80.lib.abtesting;

import java.security.SecureRandom;
import java.util.ArrayList;
import java.util.List;

/**
 * @author Olivier Perez
 */
public class ABPercentConfig {

    private final String name;
    private final String defaultValue;

    private final List<ABPercentCase> cases;

    ABPercentConfig(String name, String defaultValue) {
        this.name = name;
        this.defaultValue = defaultValue;
        this.cases = new ArrayList<>();
    }

    public void addCase(int lower, int upper, String value) {
        cases.add(new ABPercentCase(lower, upper, value));
    }

    String result(ABStore store) {
        int value = store.getInt(name, -1);
        for (ABPercentCase aCase : cases) {
            if (aCase.matches(value)) {
                return aCase.value();
            }
        }
        return defaultValue;
    }

    public void init(ABStore store) {
        if (store.getInt(name, -1) == -1) {
            SecureRandom random = new SecureRandom();
            int value = random.nextInt(99);

            store.set(name, value);
        }
    }
}

File: abtesting/src/main/java/fr/o80/lib/abtesting/ABPercentConfig.kt
package fr.o80.lib.abtesting

import java.security.SecureRandom

/**
 * @author Olivier Perez
 */
@AbTestingDsl
class ABPercentConfig internal constructor(private val name: String) {

    private val cases: MutableList<ABPercentCase> = mutableListOf()

    private lateinit var defaultValue: String
    object DefaultCase

    @AbTestingDsl
    infix fun String.isIn(range: IntRange) {
        cases.add(ABPercentCase(range.first, range.last, this))
    }

    @AbTestingDsl
    infix fun String.isFor(default: DefaultCase) {
        defaultValue = this
    }

    @AbTestingDsl

    internal fun result(store: ABStore): String {
        var value = store.getInt(name, DEFAULT_VALUE)
        value = if (value >= 0) value else init(store)

        for (aCase in cases) {
            if (aCase.matches(value)) {
                return aCase.value()
            }
        }
        return defaultValue
    }

    private fun init(store: ABStore): Int {
        val value = store.getInt(name, DEFAULT_VALUE)
        return when (value) {
            DEFAULT_VALUE -> {
                val random = SecureRandom()
                val newValue = random.nextInt(99)

                store[name] = newValue
                newValue
            }
            else -> {
                value
            }
        }
    }

    fun reset(store: ABStore) {
        store[name] = DEFAULT_VALUE
    }

    companion object {
        private const val DEFAULT_VALUE = -1
    }

}